Parsers/ASimNetworkSession/Parsers/ASimNetworkSessionCiscoASA.yml

```diff
@@ -0,0 +1,369 @@
+      let all_302013_302015_parsed = allLogs
+                | where DeviceEventClassID in ("302013","302015") and isnotempty(SourceIP)
+                | parse Message with * ":  " * " " * " " * " connection " NetworkSessionId " for " SrcInterfaceName ":" * "/" * " (" SrcNatIpAddr "/" SrcNatPortNumber:int ")" SrcUsername "to " DstInterfaceName ":" * "/" * " (" DstNatIpAddr "/" DstNatPortNumber:int ")" DstUsername
+                | extend SrcUsername = trim(@"[()]", SrcUsername),
+                        DstUsername = trim(@"[()]", DstUsername)
+                | extend 
+                        SrcUsernameType = iif(isnotempty(SrcUsername),"Windows",""),
+                        DstUsernameType = iif(isnotempty(DstUsername),"Windows",""),
+                        SimplifiedDeviceAction = DeviceAction,
+                        SessionId = NetworkSessionId;
+      let all_302013_302015_unparsed = allLogs
+                | where DeviceEventClassID in ("302013","302015") and isempty(SourceIP)
+                | parse Message with * ":  " DeviceAction " " NetworkDirection " " Protocol " connection " NetworkSessionId " for " SrcInterfaceName ":" SrcIpAddr "/" SrcPortNumber:int " (" SrcNatIpAddr "/" SrcNatPortNumber:int ")" SrcUsername "to " DstInterfaceName ":" DstIpAddr "/" DstPortNumber:int " (" DstNatIpAddr "/" DstNatPortNumber:int ")" DstUsername
+                | extend SrcUsername = trim(@"[()]", SrcUsername),
+                        DstUsername = trim(@"[()]", DstUsername),
+                        NetworkDirection = case(NetworkDirection == "inbound", "Inbound",
+                                                    NetworkDirection == "outbound", "Outbound",
+                                                    "")
+                | extend 
+                        SrcUsernameType = iif(isnotempty(SrcUsername),"Windows",""),
+                        DstUsernameType = iif(isnotempty(DstUsername),"Windows",""),
+                        SimplifiedDeviceAction = DeviceAction,
+                        SessionId = NetworkSessionId;
+      let all_302014_302016 = allLogs
+                | where DeviceEventClassID in ("302014", "302016")
+                | parse kind=regex Message with * ": " DeviceAction " " Protocol " connection " NetworkSessionId " for " SrcInterfaceName ":" SrcIpAddr "/" SrcPortNumber:int @"[(]?" SrcUsername @"[)]? to " DstInterfaceName ":" DstIpAddr "/" DstPortNumber:int @"[(]?" DstUsername1 @"[)]? [Dd]uration " NetworkDuration " bytes " DstBytes:long @"\s?" Remaining 
+                | extend reasonParsed = extract_all(@"$|\((\S+)\)|([\S+\s*]+)\sfrom\s\S+\s(\(\S+\)|\S+)$|([\S+\s*]+)\sfrom\s\S+$|([\S+\s*]+)\s\((\S+)\)$|([\S+\s*]+)$",Remaining)[0]
+                | extend 
+                            SrcUsername = trim(@"[()]", SrcUsername),
+                            DstUsername = trim(@"[()]",coalesce(reasonParsed[0], reasonParsed[2], reasonParsed[5], DstUsername1)),
```

Comment:
This sounds really error prone. Can he DstUsername really be in 3 different places in the "Remining"? Same for Reason. Also, the parse will fail if DstUsername1 is missing, want it. And if DstUsername1, what's the importance of the other fields?

Reply:
Yes, following are the different ways this Remaining string can shape:
1. <empty>
2. (domain\USER001) --> reasonParsed[0]
3. TCP FINs from OUTSIDE (domain\USER001) --> reasonParsed[1] from OUTSIDE reasonParsed[2]
4. TCP FINs from OUTSIDE --> reasonParsed[3] from OUTSIDE
5. TCP FINs (domain\USER001) --> reasonParsed[4] reasonParsed[5]

No, if DstUsername1 is not present, the remaining parsing will work fine as it is recorded as empty string.
![image](https://user-images.githubusercontent.com/10944026/187216727-18b1009b-c16b-4603-a1ea-73d9f39af262.png)